dubbo-remoting/dubbo-remoting-netty4/src/main/java/org/apache/dubbo/remoting/transport/netty4/NettyClientHandler.java

```diff
@@ -78,27 +76,35 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
         }
     }
 
-
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
         super.write(ctx, msg, promise);
-        NettyChannel channel = NettyChannel.getOrAddChannel(ctx.channel(), url, handler);
-        try {
-            // if error happens from write, mock a BAD_REQUEST response so that invoker can return immediately without
-            // waiting until timeout. FIXME: not sure if this is the right approach, but exceptionCaught doesn't work
-            // as expected.
-            if (promise.cause() != null && msg instanceof Request) {
+        final NettyChannel channel = NettyChannel.getOrAddChannel(ctx.channel(), url, handler);
+        final boolean isMsg = msg instanceof Request;
+
+        // We add listeners to make sure our out bound event is correct.
+        // If our out bound event has an error (in most cases the encoder fails),
+        // we need to have the request return directly instead of blocking the invoke process.
+        promise.addListener(future -> {
+            if (future.isSuccess()) {
+                // if our future is success, mark the future to sent.
+                try {
+                    handler.sent(channel, msg);
+                } finally {
+                    NettyChannel.removeChannelIfDisconnected(ctx.channel());
+                }
+                return;
+            }
+
+            Throwable t = future.cause();
+            if (t != null && isMsg) {
                 Request request = (Request) msg;
                 Response response = new Response(request.getId(), request.getVersion());
```

Comment:
Should we refactore this by extrating to a method to make it reusable and more readable e.g. 

```
handler.received(channel,buildErrorResponse((Request)msg);

private static Response buildErrorResponse(Request request) {
   Response response = new Response(request.getId(), request.getVersion());
   response.setStatus(Response.BAD_REQUEST);
   response.setErrorMessage(StringUtils.toString(t));
   return response;
}
```

Is the something do you feel should be good ?

Reply:
Agree with u.
This will make code clear.
I'd like fix it soon.

Reply:
thanks.